### ISSUE TITLE
유사도 측정 - 2

### ISSUE THREAD

**Issue by minhyeoky** [capstone-willson/chatbot-backend#2]
# 개선 사항

- [x]  위키 말뭉치
- [ ]  나무위키 말뭉치
- [x]  형태소 분석
- [ ]  품사 정보 주입
- [x]  bert 직접 학습
- [x]  질문 형태소 분석
- [ ]  클러스터링 후 카테고리 별 대표 벡터 생성

# Vocabulary

[Voca 문서 ](https://www.notion.so/0c3921b7d19c4995b89d3ce0b39a0d28)

# 모델 하이퍼 파라미터

    // bert_config.json
    {
      "attention_probs_dropout_prob": 0.1, 
      "directionality": "bidi", 
      "hidden_act": "gelu", 
      "hidden_dropout_prob": 0.1, 
      "hidden_size": 768, 
      "initializer_range": 0.02, 
      "intermediate_size": 3072, 
      "max_position_embeddings": 512, 
      "num_attention_heads": 12, 
      "num_hidden_layers": 12, 
      "pooler_fc_size": 768, 
      "pooler_num_attention_heads": 12, 
      "pooler_num_fc_layers": 3, 
      "pooler_size_per_head": 128, 
      "pooler_type": "first_token_transform", 
      "type_vocab_size": 2, 
      "vocab_size": 9169
    }

    // run_pretraining
    flags.DEFINE_integer(
        "max_seq_length", 128,
        "The maximum total input sequence length after WordPiece tokenization. "
        "Sequences longer than this will be truncated, and sequences shorter "
        "than this will be padded. Must match data generation.")
    flags.DEFINE_integer(
        "max_predictions_per_seq", 20,
        "Maximum number of masked LM predictions per sequence. "
        "Must match data generation.")
    flags.DEFINE_integer("train_batch_size", 32, "Total batch size for training.")
    flags.DEFINE_float("learning_rate", 5e-5, "The initial learning rate for Adam.")
    flags.DEFINE_integer("num_train_steps", 100000, "Number of training steps.")
    flags.DEFINE_integer("num_warmup_steps", 10000, "Number of warmup steps.")
    flags.DEFINE_integer("save_checkpoints_steps", 1000,
                         "How often to save the model checkpoint.")

![image](https://user-images.githubusercontent.com/12870549/55287223-182f2280-53e1-11e9-9212-845b2183a2b1.png)

- Loss: 2.2829
- step: 100000

---

# 유사도 분석 테스트

    TEST3 = {
        'vocab_file': os.path.join(BASE_DIR, '../data/vocab_mecab+khaiii_noBPE_5888'),
        'max_seq_length': 128,
        'max_query_length': 128,
        'model_path': os.path.join(BASE_DIR, '../data/output_dir/model.ckpt-100000'),
        'bert_json': os.path.join(BASE_DIR, '../data/bert_config.json'),
        'categories': ['셔틀', '밥', '잡담', '학사행정', '검색'],
        'feature_layers': -1
    }
![image](https://user-images.githubusercontent.com/12870549/55287238-3b59d200-53e1-11e9-921f-b4dcaf516f16.png)

Q: 통학버스 타는 법?

Top1과의 맨하탄 거리: 123.22

Top2와 거리 25, 
![image](https://user-images.githubusercontent.com/12870549/55287242-46acfd80-53e1-11e9-9403-9bf80d9f61b3.png)

Q: 캠퍼스 전과 할 시 일반 전과와의 평가 점수 차이점?

Top1과의 맨하탄 거리: 124.81

Top2와 거리 50정도로 나쁘지 않음, 

![image](https://user-images.githubusercontent.com/12870549/55287245-4e6ca200-53e1-11e9-83aa-19cbd4e8a7a9.png)

Q: 기숙사 어떻게 들어가?

정답과의 거리 262, Voca에 

'기숙사' 라는 단어가 없었고, '기','사'는 있었지만 '숙'은 존재 하지 않았기 때문에

기숙사가 [UNK]토큰화 되었을 것, 아무런 연관성을 찾지 못 한 것으로 보인다.

# 결론

1. 유사도 분석
    - 전체 문장의 길이 짧아짐 → 유사도 구별 능력 향상
    - 형태소 분석
    - 이전 보다 빠른 속도 (3초대에서 1초 미만으로)
2. 형태소 분석
    - Mecab, khaiii모두 사용
    - khaiii의 느린속도 → memoization → 속도 향상(mecab 단일 속도까지)
3. Bert 훈련
    - SeqLength 128 → 빠른 속도
    - 10만 스텝 약 14시간 소모(1080Ti)
    - 훈련데이터 생성에 약 4시간 소모(20GB)

# 개선 할 점

1. 유사도 분석
    - 감지하지 못하는 단어들 해결
    - 비슷한 질문들 클러스터링
    - Bert의 마지막 sequence layer 를 pooling하지 말고 중간에 있는 레이어를 사용 해 보기(bert-as-service 참고)
2. 형태소 분석
3. Bert훈련
    - SeqLength 384 로 늘려서 재훈련
    - 20만스텝  및 그 이후까지 훈련 후 다시 테스트
    - 나무위키 말뭉치를 사용한 모델 훈련 및 합쳐보기
    - 품사정보 주입(실험적)
4. 훈련된 Bert기반으로 NER(개체명인식) 모델 훈련 및 테스트
